### ISSUE TITLE
Better APIs for extending integrate() + documentation

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#23624]
`integrate()` is currently extensible in many different ways, but not all of these are documented very well. I was going to write something about this in https://github.com/sympy/sympy/pull/23488 but it's clearly complicated enough that it can use its own guide. Also, many of the APIs for extending aren't particularly good and should perhaps be improved before we make them officially "public" APIs.

So far, there are at least 4 ways to extend `integrate`:

- `_eval_Integral`. This bypasses `integrate` completely. But it has a few issues:
  - It passes the arguments from `Integral` unmodified, meaning you have to parse
    the symbols.
  - It only works if the function in question is the exact function passed to
    `integrate()`, e.g., it would do nothing for `integrate(x*f(x**2))`, because
    it's only used at the `Integrate` top-level. The various algorithms don't use it.
  
  Consequently, I don't think it should ever be used, except for custom objects for which the current integration routines would know nothing about (e.g., `Quaternion._eval_Integral` is defined)
  
  Perhaps a stripped down `_eval_integral` that prepreocesses the symbols (similar to how `_eval_derivative` works) would be more useful. There actually is already such a method on some classes, but it's not clear when exactly `integrate` uses it. If you define a `Function` subclass with `_eval_integral` (lowercase), it doesn't seem to be called. But it is defined on some classes like `Piecewise`. 

  However, in general, this approach isn't very good, because you don't really tell most integration algorithms how to integrate something like this. Instead you tell it about some other identities the function satisfies and they figure out how to integrate expressions containing the function from that. An exception might be manualintegrate with things like u-substitution and integration by parts (see below).

- heurisch can be extended by defining derivatives. However, the derivatives must be defined in such a way that they are in terms of the original function. This is sometimes annoying as it requires writing a derivative in a non-simple way. For example, for the `versin(x) = 1 - cos(x)` function I am using in the custom functions guide, if you do the obvious thing and define `versin(x).diff(x) == sin(x)`, it doesn't work. Instead, you have to define `versin(x).diff(x) = 1 - versin(x - pi/2)`. And furthermore, you have to add simplification of `versin(x - pi) = -versin(x)` to the evaluator or else heurisch won't recognize it as being in terms of the same function. If you do this, the integration does work:

  ```py
  >>> from sympy.functions.elementary.trigonometric import _peeloff_pi
  >>> class versin(Function):
  ...     @classmethod
  ...     def eval(cls, x):
  ...         rest, pi_mult = _peeloff_pi(x)
  ...         if pi_mult and isinstance(pi_mult, Integer):
  ...             return versin(rest)*(-1)**pi_mult
  ...
  ...     def fdiff(self, argindex=1):
  ...         return -versin(self.args[0] - pi/2) + 1
  >>> from sympy import integrate
  >>> integrate(versin(x), x)
  -x + versin(x - pi/2)
  >>> integrate(versin(x)*x, x)
  -x**2/2 + x*versin(x - pi/2) - x + versin(x)
  ```

  So ideally there should be a separate method for defining the derivative just for heurisch. See https://github.com/sympy/sympy/issues/4300 and various cross-linked issues. 

- The Meijer G-Function algorithm can be extended by extending the hyperexpand tables (see https://docs.sympy.org/dev/modules/simplify/hyperexpand.html#extending-the-hypergeometric-tables). However, this is quite mathematically complicated, although it works quite well when applied appropriately. It's also not completely clear from that guide that it ought to be supported for user-defined functions and not just functions defined within SymPy. 

- manualintegrate should be extensible, in principle, in the same ways it is extended in-library. However, nothing is documented, so we should make sure we like the API before making it public, and then document it.

So what we need to do is:

- Settle on more "public" APIs for defining integration. Ideally there should be methods that can be defined on a function itself, although in many cases, extending the algorithm itself makes more sense, because the appropriate rules might apply to larger expressions than just the function itself. The APIs should also try to avoid redundancy. For example, there shouldn't be more than one way to define what integrate(f(x)) is. Right now, you can define it separately with `_eval_Integral` and in a `manualintegrate` rule. The methods should be equally applicable to functions defined within SymPy and user-defined functions. 

- Document these, ideally in a separate user guide similar to the custom functions guide [I am writing](https://github.com/sympy/sympy/pull/23488).

CC @eagleoflqj, curious if you have any thoughts on this, since it is related to some of the work you have been doing recently.

**Comment by eagleoflqj** [comment]
> However, in general, this approach isn't very good, because you don't really tell most integration algorithms how to integrate something like this.

Yes. I'm not sure how much we can benefit from defining `_eval_integral` for specific function classes, as it's not as powerful as defining derivative to be used in chain rule.

> manualintegrate should be extensible, in principle, in the same ways it is extended in-library. However, nothing is documented, so we should make sure we like the API before making it public, and then document it.

`manualintegrate` has a great design framework. Though it does need better documentation. When I started to work on it, I found it not straightforward to understand, and the untyped `namedtuple`s just added more difficulty. I think we should use conventional classes with inheritance instead, and provide docstring for every rule class.

For other aspects of this issue I'm not able to discuss more due to limited code base understanding.

**Comment by asmeurer** [comment]
I was mostly looking for your views on manualintegrate since that's what you've worked on. IIRC it was already refactored somewhat recently, so we should take a look at what was done there before refactoring it again.
